File: dotnet/CardsProject/Services/Cards/CardService.cs
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;
using CardsProject.Controllers.Cards;
using CardsProject.Domain.Cards;
using CardsProject.Infrastructure;

namespace CardsProject.Services.Cards;

public class CardService
{
    private readonly AppDbContext _db;

    public CardService(AppDbContext db) => _db = db;

    private static void ValidateEntity(object entity)
    {
        var ctx = new ValidationContext(entity);
        Validator.ValidateObject(entity, ctx, validateAllProperties: true);
    }

    public async Task<List<Card>> GetAllAsync()
        => await _db.Cards.AsNoTracking().ToListAsync();

    public async Task<Card?> GetByIdAsync(int id)
        => await _db.Cards.FindAsync(id);

    public async Task<Card> CreateAsync(CardDto dto)
    {
        var entity = new Card();
        if (dto.Name is not null) entity.Name = dto.Name;
        if (dto.CardType is not null && Enum.TryParse<CardCardTypeType>(dto.CardType, out var cardTypeVal)) entity.CardType = cardTypeVal;
        if (dto.Rarity is not null && Enum.TryParse<CardRarityType>(dto.Rarity, out var rarityVal)) entity.Rarity = rarityVal;
        if (dto.ManaCost is not null) entity.ManaCost = dto.ManaCost.Value;
        if (dto.ManaColors is not null && Enum.TryParse<CardManaColorsType>(dto.ManaColors, out var manaColorsVal)) entity.ManaColors = manaColorsVal;
        if (dto.Attack is not null) entity.Attack = dto.Attack.Value;
        if (dto.Defense is not null) entity.Defense = dto.Defense.Value;
        if (dto.Loyalty is not null) entity.Loyalty = dto.Loyalty.Value;
        if (dto.Description is not null) entity.Description = dto.Description;
        if (dto.FlavorText is not null) entity.FlavorText = dto.FlavorText;
        if (dto.ImageUrl is not null) entity.ImageUrl = dto.ImageUrl;
        if (dto.ArtistName is not null) entity.ArtistName = dto.ArtistName;
        if (dto.LegalFormats is not null && Enum.TryParse<CardLegalFormatsType>(dto.LegalFormats, out var legalFormatsVal)) entity.LegalFormats = legalFormatsVal;
        if (dto.IsBanned is not null) entity.IsBanned = dto.IsBanned.Value;
        if (dto.IsRestricted is not null) entity.IsRestricted = dto.IsRestricted.Value;
        if (dto.PowerLevel is not null) entity.PowerLevel = dto.PowerLevel.Value;
        if (dto.SetId is not null) entity.SetId = dto.SetId;
        Validate(entity);
        ValidateEntity(entity);
        _db.Cards.Add(entity);
        await _db.SaveChangesAsync();
        return entity;
    }

    public async Task<Card?> UpdateAsync(int id, CardDto dto)
    {
        var entity = await _db.Cards.FindAsync(id);
        if (entity is null) return null;
        if (dto.Name is not null) entity.Name = dto.Name;
        if (dto.CardType is not null && Enum.TryParse<CardCardTypeType>(dto.CardType, out var cardTypeVal)) entity.CardType = cardTypeVal;
        if (dto.Rarity is not null && Enum.TryParse<CardRarityType>(dto.Rarity, out var rarityVal)) entity.Rarity = rarityVal;
        if (dto.ManaCost is not null) entity.ManaCost = dto.ManaCost.Value;
        if (dto.ManaColors is not null && Enum.TryParse<CardManaColorsType>(dto.ManaColors, out var manaColorsVal)) entity.ManaColors = manaColorsVal;
        if (dto.Attack is not null) entity.Attack = dto.Attack.Value;
        if (dto.Defense is not null) entity.Defense = dto.Defense.Value;
        if (dto.Loyalty is not null) entity.Loyalty = dto.Loyalty.Value;
        if (dto.Description is not null) entity.Description = dto.Description;
        if (dto.FlavorText is not null) entity.FlavorText = dto.FlavorText;
        if (dto.ImageUrl is not null) entity.ImageUrl = dto.ImageUrl;
        if (dto.ArtistName is not null) entity.ArtistName = dto.ArtistName;
        if (dto.LegalFormats is not null && Enum.TryParse<CardLegalFormatsType>(dto.LegalFormats, out var legalFormatsVal)) entity.LegalFormats = legalFormatsVal;
        if (dto.IsBanned is not null) entity.IsBanned = dto.IsBanned.Value;
        if (dto.IsRestricted is not null) entity.IsRestricted = dto.IsRestricted.Value;
        if (dto.PowerLevel is not null) entity.PowerLevel = dto.PowerLevel.Value;
        if (dto.SetId is not null) entity.SetId = dto.SetId;
        Validate(entity);
        ValidateEntity(entity);
        await _db.SaveChangesAsync();
        return entity;
    }

    public async Task<bool> DeleteAsync(int id)
    {
        var entity = await _db.Cards.FindAsync(id);
        if (entity is null) return false;
        _db.Cards.Remove(entity);
        await _db.SaveChangesAsync();
        return true;
    }

    public async System.Threading.Tasks.Task<bool> BanAsync(int id)
    {
        var entity = await _db.Cards.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("Card not found: " + id);
        entity.Ban();
        await _db.SaveChangesAsync();
        return true;
    }
    public async System.Threading.Tasks.Task<bool> UnbanAsync(int id)
    {
        var entity = await _db.Cards.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("Card not found: " + id);
        entity.Unban();
        await _db.SaveChangesAsync();
        return true;
    }
    public async System.Threading.Tasks.Task<bool> RestrictAsync(int id)
    {
        var entity = await _db.Cards.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("Card not found: " + id);
        entity.Restrict();
        await _db.SaveChangesAsync();
        return true;
    }
    public async System.Threading.Tasks.Task<bool> UnrestrictAsync(int id)
    {
        var entity = await _db.Cards.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("Card not found: " + id);
        entity.Unrestrict();
        await _db.SaveChangesAsync();
        return true;
    }
    public async System.Threading.Tasks.Task<decimal> CalculateValueAsync(int id)
    {
        var entity = await _db.Cards.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("Card not found: " + id);
        var result = entity.CalculateValue();
        await _db.SaveChangesAsync();
        return result;
    }
    public async System.Threading.Tasks.Task<decimal> ApplyRarityBonusAsync(int id, int multiplier)
    {
        var entity = await _db.Cards.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("Card not found: " + id);
        var result = entity.ApplyRarityBonus(multiplier);
        await _db.SaveChangesAsync();
        return result;
    }
    public async System.Threading.Tasks.Task<bool> IsLegalInFormatAsync(int id, string format)
    {
        var entity = await _db.Cards.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("Card not found: " + id);
        var result = entity.IsLegalInFormat(format);
        await _db.SaveChangesAsync();
        return result;
    }
    public void Validate(Card entity)
    {
        if (entity.CardType == CardCardTypeType.Creature && !(entity.Attack != null && entity.Defense != null)) throw new InvalidOperationException("Creature card must have attack and defense");
        if (entity.CardType == CardCardTypeType.Planeswalker && entity.Loyalty == null) throw new InvalidOperationException("Planeswalker card must have loyalty");
        if (entity.CardType != CardCardTypeType.Planeswalker && entity.Loyalty != null) throw new InvalidOperationException("Only Planeswalker cards can have loyalty");
        if (entity.IsBanned == true && true) throw new InvalidOperationException("banned_card_not_in_legal_formats");
    }
}

File: spring/src/main/java/cardsproject/service/cards/CardService.java
package cardsproject.service.cards;

import cardsproject.domain.cards.Card;
import cardsproject.repository.cards.CardRepository;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.Optional;
import java.util.Objects;
import cardsproject.domain.cards.CardCardTypeType;
import cardsproject.domain.cards.CardRarityType;
import cardsproject.domain.cards.CardManaColorsType;
import cardsproject.domain.cards.CardLegalFormatsType;

@Service
public class CardService {

    private final CardRepository repository;

    public CardService(CardRepository repository) {
        this.repository = repository;
    }

    public List<Card> findAll() {
        return repository.findAll();
    }

    public List<Card> search(String q) {
        if (q == null || q.isBlank()) return repository.findAll();
        return repository.findAll().stream()
            .filter(e -> (e.getName() != null && e.getName().toLowerCase().contains(q.toLowerCase())) || (e.getArtistName() != null && e.getArtistName().toLowerCase().contains(q.toLowerCase())))
            .collect(java.util.stream.Collectors.toList());
    }

    public Optional<Card> findById(Long id) {
        return repository.findById(id);
    }

    public Card save(Card entity) {
        validate(entity);
        return repository.save(entity);
    }

    public void delete(Long id) {
        repository.deleteById(id);
    }

    public void applyPatch(Card entity, java.util.Map<String, Object> patch) {
        if (patch.containsKey("name") && patch.get("name") != null) entity.setName(patch.get("name").toString());
        if (patch.containsKey("cardType")) entity.setCardType(CardCardTypeType.valueOf(patch.get("cardType").toString()));
        if (patch.containsKey("rarity")) entity.setRarity(CardRarityType.valueOf(patch.get("rarity").toString()));
        if (patch.containsKey("manaCost") && patch.get("manaCost") != null) entity.setManaCost(Integer.valueOf(patch.get("manaCost").toString()));
        if (patch.containsKey("manaColors")) entity.setManaColors(CardManaColorsType.valueOf(patch.get("manaColors").toString()));
        if (patch.containsKey("attack") && patch.get("attack") != null) entity.setAttack(Integer.valueOf(patch.get("attack").toString()));
        if (patch.containsKey("defense") && patch.get("defense") != null) entity.setDefense(Integer.valueOf(patch.get("defense").toString()));
        if (patch.containsKey("loyalty") && patch.get("loyalty") != null) entity.setLoyalty(Integer.valueOf(patch.get("loyalty").toString()));
        if (patch.containsKey("description") && patch.get("description") != null) entity.setDescription(patch.get("description").toString());
        if (patch.containsKey("flavorText") && patch.get("flavorText") != null) entity.setFlavorText(patch.get("flavorText").toString());
        if (patch.containsKey("imageUrl") && patch.get("imageUrl") != null) entity.setImageUrl(patch.get("imageUrl").toString());
        if (patch.containsKey("artistName") && patch.get("artistName") != null) entity.setArtistName(patch.get("artistName").toString());
        if (patch.containsKey("legalFormats")) entity.setLegalFormats(CardLegalFormatsType.valueOf(patch.get("legalFormats").toString()));
        if (patch.containsKey("isBanned") && patch.get("isBanned") != null) entity.setIsBanned(Boolean.valueOf(patch.get("isBanned").toString()));
        if (patch.containsKey("isRestricted") && patch.get("isRestricted") != null) entity.setIsRestricted(Boolean.valueOf(patch.get("isRestricted").toString()));
        if (patch.containsKey("powerLevel") && patch.get("powerLevel") != null) entity.setPowerLevel(Integer.valueOf(patch.get("powerLevel").toString()));
        if (patch.containsKey("setId") && patch.get("setId") != null) entity.setSetId(Long.valueOf(patch.get("setId").toString()));
    }
    private void validate(Card entity) {
        if (CardCardTypeType.CREATURE.equals(entity.getCardType()) && !(entity.getAttack() != null && entity.getDefense() != null)) throw new IllegalStateException("Creature card must have attack and defense");
        if (CardCardTypeType.PLANESWALKER.equals(entity.getCardType()) && entity.getLoyalty() == null) throw new IllegalStateException("Planeswalker card must have loyalty");
        if (CardCardTypeType.LAND.equals(entity.getCardType()) && !(Objects.equals(entity.getManaCost(), 0))) throw new IllegalStateException("Land card must have zero mana cost");
        if (!CardCardTypeType.PLANESWALKER.equals(entity.getCardType()) && entity.getLoyalty() != null) throw new IllegalStateException("Only Planeswalker cards can have loyalty");
        if (Boolean.TRUE.equals(entity.getIsBanned()) && true) throw new IllegalStateException("banned_card_not_in_legal_formats");
    }

    public void ban(Long id) {
        Card entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Card not found: " + id));
        entity.ban();
        repository.save(entity);
    }

    public void unban(Long id) {
        Card entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Card not found: " + id));
        entity.unban();
        repository.save(entity);
    }

    public void restrict(Long id) {
        Card entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Card not found: " + id));
        entity.restrict();
        repository.save(entity);
    }

    public void unrestrict(Long id) {
        Card entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Card not found: " + id));
        entity.unrestrict();
        repository.save(entity);
    }

    public java.math.BigDecimal calculateValue(Long id) {
        Card entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Card not found: " + id));
        java.math.BigDecimal result = entity.calculateValue();
        repository.save(entity);
        return result;
    }

    public java.math.BigDecimal applyRarityBonus(Long id, Integer multiplier) {
        Card entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Card not found: " + id));
        java.math.BigDecimal result = entity.applyRarityBonus(multiplier);
        repository.save(entity);
        return result;
    }

    public Boolean isLegalInFormat(Long id, String format) {
        Card entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Card not found: " + id));
        Boolean result = entity.isLegalInFormat(format);
        repository.save(entity);
        return result;
    }
}
